### ISSUE TITLE
Conflicting requirements in OTLP JSON field names

### ISSUE THREAD

**Issue by tigrannajaryan** [open-telemetry/opentelemetry-specification#3365]
## Problem

[JSON Protobuf Encoding](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#json-protobuf-encoding) section in OTLP protocol specification says:

>The trace_id and span_id byte arrays are represented as ...

Note that `trace_id` and `span_id` fields use lower snake case, with underscores, using original field names from the [proto](https://github.com/open-telemetry/opentelemetry-proto/blob/7aa439cb0ba78afbd009d83f32fdda6c128e0342/opentelemetry/proto/logs/v1/logs.proto#L177).

However, in the same section of the spec we also have this:

>The keys of JSON objects are field names converted to lowerCamelCase. **Original field names are not valid to use as keys for JSON objects**. For example, this is a valid JSON representation of a Resource: { "attributes": {...}, "droppedAttributesCount": 123 }, and this is NOT a valid representation: { "attributes": {...}, "dropped_attributes_count": 123 }.

These 2 clauses conflict with each other. Which one is true? Should we use lower camel case, i.e. `traceId`/`spanId` or the original names `trace_id`/`span_id`?

We ended up in this broken situation after [this PR](https://github.com/open-telemetry/opentelemetry-specification/pull/2829/files), where we decided that lowerCamelCase names only must be used.

The spec contradicts with itself and it needs fixing. 

## Possible solutions

I see 2 possible ways forward:

### Rename `trace_id`/`span_id` to `traceId`/`spanId`

We will restore consistency of the spec and eliminate contradictory statements.

However this is very disruptive and breaks interoperability of old and new OTLP implementations. I believe this breaks our stability promises and should not be allowed.

### Allow `trace_id`/`span_id` names as an exception

We can explicitly call out in the spec that the names of these 2 fields are an exception from the rule about lowerCamelCase. This is the least disruptive option. Implementations don't need to change, everything continues working as is. We will end up with a bit ugliness and inconsistency in the OTLP JSON spec.

I think this should be our choice.

If anyone can think of a better solution please comment.

**Comment by jack-berg** [comment]
> The trace_id and span_id byte arrays are represented as [case-insensitive hex-encoded strings](https://tools.ietf.org/html/rfc4648#section-8); they are not base64-encoded as is defined in the standard [Protobuf JSON Mapping](https://developers.google.com/protocol-buffers/docs/proto3#json). Hex encoding is used for trace_id and span_id fields in all OTLP Protobuf messages, e.g., the Span, Link, LogRecord, etc. messages. For example, the trace_id field in a Span can be represented like this: { "trace_id": "5B8EFFF798038103D269B633813FC60C", ... }

The reference to snake case here seems like a typo. This clause is referencing the encoding of the strings, and is not concerns with the casing of keys. I believe the snake case was probably in reference to the field names in the [proto definition](https://github.com/open-telemetry/opentelemetry-proto/blob/7aa439cb0ba78afbd009d83f32fdda6c128e0342/opentelemetry/proto/trace/v1/trace.proto#L87-L95). 

I'm in favor of the first option, but would like to understand more about which components would be disrupted by this.

**Comment by tigrannajaryan** [comment]
>I'm in favor of the first option, but would like to understand more about which components would be disrupted by this.

Any existing sender implementation of OTLP JSON which is compliant with the current spec will be no longer able to send spans to any other receiver implementation of OTLP JSON that is compliant with the modified Otel spec (and vice versa). 

This is a catastrophic failure mode for me, the worst kind of a failure to deliver on our stability promise since the only way to avoid being affected by it is co-ordinated updating of all participants (producers and consumers) or a distributed network, which in any non-trivial deployment is either hard or impossible to do.

**Comment by jack-berg** [comment]
> Any existing sender implementation of OTLP JSON which is compliant with the current spec will be no longer able to send spans to any other receiver implementation of OTLP JSON that is compliant with the modified Otel spec (and vice versa).

Doesn't that assume that a sender producing `trace_id` / `span_id` is in fact compliant? As you point out, the language is contradictory. You can make the case that the later bullet point means that sending `trace_id` / `span_id` instead of `traceId` / `spanId` means the sender is not compliant.

**Comment by tigrannajaryan** [comment]
You are right. It is a matter of which of the contradicting clauses we think was the overriding one and I don't think there is a clear answer to that.

**Comment by punya** [comment]
Note that the [example JSON files](https://github.com/open-telemetry/opentelemetry-proto/blob/449a1b0219a699a60d9cbc3899c68cb1ec8c1139/examples/trace.json) in the repository actually use *both* types of casing in the same file!
* Snake case: `resource_spans`, `scope_spans`
* Camel case: `stringValue`, `parentSpanId`
